### PR TITLE
[KnownBits] Make `avg{Ceil,Floor}S` optimal

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[KnownBits] Make `avg{Ceil,Floor}S` optimal**

All we where missing was the signbit if we knew the incoming signbit
of either LHS or RHS.

Since the base addition in the average is with an extra bit width it
cannot overflow, we figure out the result sign based on the magnitude
of the input. If the negative component has a larger magnitude the
result is negative and vice versa for the positive case.

### CODE REVIEW

File: llvm/lib/Support/KnownBits.cpp [hunk 1]
```diff
@@ -766,12 +775,44 @@ KnownBits KnownBits::usub_sat(const KnownBits &LHS, const KnownBits &RHS) {
...
-      computeForAddCarry(LHS, RHS, /*CarryZero*/ !IsCeil, /*CarryOne*/ IsCeil);
-  LHS = LHS.extractBits(BitWidth, 1);
-  return LHS;
+  KnownBits ExtLHS = IsSigned ? LHS.sext(BitWidth + 1) : LHS.zext(BitWidth + 1);
+  KnownBits ExtRHS = IsSigned ? RHS.sext(BitWidth + 1) : RHS.zext(BitWidth + 1);
+  KnownBits Res = computeForAddCarry(ExtLHS, ExtRHS, /*CarryZero*/ !IsCeil,
+                                     /*CarryOne*/ IsCeil);
+  Res = Res.extractBits(BitWidth, 1);
+
+  auto SignBitKnown = [BitWidth](KnownBits KB) {
```
Comment (dtcxzyw) [inline llvm/lib/Support/KnownBits.cpp:784]: We already have `KnownBits::isSignUnknown`.

File: llvm/lib/Support/KnownBits.cpp [hunk 2]
```diff
@@ -766,12 +775,44 @@ KnownBits KnownBits::usub_sat(const KnownBits &LHS, const KnownBits &RHS) {
...
-  LHS = IsSigned ? LHS.sext(BitWidth + 1) : LHS.zext(BitWidth + 1);
-  RHS = IsSigned ? RHS.sext(BitWidth + 1) : RHS.zext(BitWidth + 1);
-  LHS =
-      computeForAddCarry(LHS, RHS, /*CarryZero*/ !IsCeil, /*CarryOne*/ IsCeil);
-  LHS = LHS.extractBits(BitWidth, 1);
-  return LHS;
+  KnownBits ExtLHS = IsSigned ? LHS.sext(BitWidth + 1) : LHS.zext(BitWidth + 1);
+  KnownBits ExtRHS = IsSigned ? RHS.sext(BitWidth + 1) : RHS.zext(BitWidth + 1);
+  KnownBits Res = computeForAddCarry(ExtLHS, ExtRHS, /*CarryZero*/ !IsCeil,
+                                     /*CarryOne*/ IsCeil);
```
Comment (dtcxzyw) [inline llvm/lib/Support/KnownBits.cpp:781]:
```suggestion
  KnownBits Res = computeForAddCarry(ExtLHS, ExtRHS, /*CarryZero=*/!IsCeil,
                                     /*CarryOne=*/IsCeil);
```

File: llvm/lib/Support/KnownBits.cpp [hunk 3]
```diff
@@ -656,52 +706,8 @@ static KnownBits computeForSatAddSub(bool Add, bool Signed,
...
-    if (Add) {
-      Res.One |= Mask;
-      Res.Zero &= ~Mask;
-    } else {
-      Res.Zero |= Mask;
-      Res.One &= ~Mask;
-    }
-  }
+  KnownBits Res = KnownBits::computeForAddSub(Add, /*NSW=*/Signed,
+                                              /*NUW*/ !Signed, LHS, RHS);
```
Comment (dtcxzyw) [inline llvm/lib/Support/KnownBits.cpp:710]:
```suggestion
                                              /*NUW=*/!Signed, LHS, RHS);
```

File: llvm/lib/Support/KnownBits.cpp [hunk 4]
```diff
@@ -714,8 +720,9 @@ static KnownBits computeForSatAddSub(bool Add, bool Signed,
     APInt C;
     if (Signed) {
       // sadd.sat / ssub.sat
-      assert(SignBitKnown(LHS) &&
-             "We somehow know overflow without knowing input sign");
+      assert(LHS.isNegative() ||
+             LHS.isNonNegative() &&
```
Comment (dtcxzyw) [inline llvm/lib/Support/KnownBits.cpp:724]:
```suggestion
      assert(!LHS.isSignUnknown() &&
```

File: llvm/lib/Support/KnownBits.cpp [hunk 5]
```diff
@@ -610,28 +610,78 @@ static KnownBits computeForSatAddSub(bool Add, bool Signed,
...
+      Overflow = false;
+    else {
+      // Check if we may overflow. If we can't rule out overflow then check if
+      // we can rule out a direction at least.
+      KnownBits UnsignedLHS = LHS;
+      KnownBits UnsignedRHS = RHS;
+      UnsignedLHS.One.clearSignBit();
+      UnsignedLHS.Zero.setSignBit();
+      UnsignedRHS.One.clearSignBit();
+      UnsignedRHS.Zero.setSignBit();
```
Comment (RKSimon) [inline llvm/lib/Support/KnownBits.cpp:638]: I was going to suggest to use KnownBits.makeNonNegative() - but for some reason it doesn't clear One.clearSignBit()?

Comment (goldsteinn) [reply llvm/lib/Support/KnownBits.cpp:638]: Its been an annoyance for a file tbh, ill add a patch with `forceNegative` and `forceNonNegative` API

Comment (goldsteinn) [reply llvm/lib/Support/KnownBits.cpp:638]: See: https://github.com/llvm/llvm-project/pull/110389

File: llvm/unittests/Support/KnownBitsTest.cpp [hunk 6]
```diff
@@ -306,14 +306,14 @@ TEST(KnownBitsTest, BinaryExhaustive) {
...
-      /*CheckOptimality=*/false);
+      /*CheckOptimality=*/true);
   testBinaryOpExhaustive(
       "sub",
       [](const KnownBits &Known1, const KnownBits &Known2) {
         return KnownBits::sub(Known1, Known2);
       },
       [](const APInt &N1, const APInt &N2) { return N1 - N2; },
-      /*CheckOptimality=*/false);
+      /*CheckOptimality=*/true);
```
Comment (RKSimon) [inline llvm/unittests/Support/KnownBitsTest.cpp:316]: Is this related to the patch?

**Comment by jayfoad** [comment]
There's a lot to understand here. Could you split it into two patches, for avg and add/sub sat? Or are they inextricably intertwined?

**Comment by goldsteinn** [comment]
> There's a lot to understand here. Could you split it into two patches, for avg and add/sub sat? Or are they inextricably intertwined?

They are not. Will split.

File: llvm/unittests/Support/KnownBitsTest.cpp (hunk 6 above)
Comment (goldsteinn) [reply llvm/unittests/Support/KnownBitsTest.cpp]: No, this was just motivated by the fact that I noticed we had incorrectly labelled a lot of the tests. I pushed an NFC change to correctly label them.

**Comment by goldsteinn** [comment]
> There's a lot to understand here. Could you split it into two patches, for avg and add/sub sat? Or are they inextricably intertwined?

Split, just avg for now. Hadn't realized there was an already assigned outstanding issue to deal with addsub, so I'm going to give a few days for the original assignee to take over the addsub patches if they want to. If they don't ill post the addsub stuff in a few days.

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by goldsteinn** [comment]
@jayfoad has a much cleaner impl
